# Week 4 (2/22-2/28)

## [Pandas](../Tools/Pandas/pandas_toc.rst)

* Series and DataFrames
* Selecting data
* Sorting 
* Aggregations

## Resources

### 1. Planets data

In [12]:
planets = ["Mercury", "Venus", "Earth", "Mars", "Jupyter", "Saturn", "Uranus", "Neptune"]
diameters = [4879, 12104, 12756, 6792, 142984, 120536, 51118, 49528]
temperatures = [167, 464, 15, -65, -110, -140, -195, -200]
gravity = [3.7, 8.9, 9.8, 3.7, 23.1, 9.0, 8.7, 11.0]

## Exercises

All exercises below use data on passengers of Titanic. A DataFrame with this data can be created as follows:

In [13]:
import seaborn as sns
df = sns.load_dataset("titanic")
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

**Note.** This is partial data on Titanic passengers. It consists of 891 records, while the total number of Titanic passengers was over 2000. 

### Exercise 1

Create a DataFrame with records of all males.

**Check:** The DataFrame should consist of 577 rows.

In [82]:
# "nbsphinx": "hidden"

df[df["sex"]=="male"]

survived  pclass   sex   age  sibsp  parch     fare embarked   class  \
0           0       3  male  22.0      1      0   7.2500        S   Third   
4           0       3  male  35.0      0      0   8.0500        S   Third   
5           0       3  male   NaN      0      0   8.4583        Q   Third   
6           0       1  male  54.0      0      0  51.8625        S   First   
7           0       3  male   2.0      3      1  21.0750        S   Third   
..        ...     ...   ...   ...    ...    ...      ...      ...     ...   
883         0       2  male  28.0      0      0  10.5000        S  Second   
884         0       3  male  25.0      0      0   7.0500        S   Third   
886         0       2  male  27.0      0      0  13.0000        S  Second   
889         1       1  male  26.0      0      0  30.0000        C   First   
890         0       3  male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
4      man        True  NaN  Southampton    no   True  
5      man        True  NaN   Queenstown    no   True  
6      man        True    E  Southampton    no   True  
7    child       False  NaN  Southampton    no  False  
..     ...         ...  ...          ...   ...    ...  
883    man        True  NaN  Southampton    no   True  
884    man        True  NaN  Southampton    no   True  
886    man        True  NaN  Southampton    no   True  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[577 rows x 15 columns]

### Exercise 2

Create a DataFrame with records of all males who survived the sinking of Titanic. 

**Check:** The DataFrame should consist of 109 rows.

In [81]:
# "nbsphinx": "hidden"

df[(df["sex"]=="male") & (df["survived"] == 1)]

survived  pclass   sex   age  sibsp  parch     fare embarked   class  \
17          1       2  male   NaN      0      0  13.0000        S  Second   
21          1       2  male  34.0      0      0  13.0000        S  Second   
23          1       1  male  28.0      0      0  35.5000        S   First   
36          1       3  male   NaN      0      0   7.2292        C   Third   
55          1       1  male   NaN      0      0  35.5000        S   First   
..        ...     ...   ...   ...    ...    ...      ...      ...     ...   
838         1       3  male  32.0      0      0  56.4958        S   Third   
839         1       1  male   NaN      0      0  29.7000        C   First   
857         1       1  male  51.0      0      0  26.5500        S   First   
869         1       3  male   4.0      1      1  11.1333        S   Third   
889         1       1  male  26.0      0      0  30.0000        C   First   

       who  adult_male deck  embark_town alive  alone  
17     man        True  NaN  Southampton   yes   True  
21     man        True    D  Southampton   yes   True  
23     man        True    A  Southampton   yes   True  
36     man        True  NaN    Cherbourg   yes   True  
55     man        True    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
838    man        True  NaN  Southampton   yes   True  
839    man        True    C    Cherbourg   yes   True  
857    man        True    E  Southampton   yes   True  
869  child       False  NaN  Southampton   yes  False  
889    man        True    C    Cherbourg   yes   True  

[109 rows x 15 columns]

### Exercise 3

Find the age of the oldest Titanic passenger.

**Check:** You should get 80.

In [68]:
# "nbsphinx": "hidden"

df["age"].max()

80.0

### Exercise 4

Get the record of the oldest passenger. 

**Check:** The passenger was a male, embarked in Southampton, traveled in the first class and survived. 

In [69]:
# "nbsphinx": "hidden"

df.loc[df["age"].idxmax()]

survived                 1
pclass                   1
sex                   male
age                     80
sibsp                    0
parch                    0
fare                    30
embarked                 S
class                First
who                    man
adult_male            True
deck                     A
embark_town    Southampton
alive                  yes
alone                 True
Name: 630, dtype: object

### Exercise 5

Create a DataFrame with records of the 5 oldest females. 

**Check:** The ages of the women should be: 63, 63, 62, 60, and 58. 

In [70]:
# "nbsphinx": "hidden"

df2 = df.sort_values(by = "age", ascending = False)
df2[df2["sex"] == "female"].head(5)

survived  pclass     sex   age  sibsp  parch      fare embarked  class  \
275         1       1  female  63.0      1      0   77.9583        S  First   
483         1       3  female  63.0      0      0    9.5875        S  Third   
829         1       1  female  62.0      0      0   80.0000      NaN  First   
366         1       1  female  60.0      1      0   75.2500        C  First   
268         1       1  female  58.0      0      1  153.4625        S  First   

       who  adult_male deck  embark_town alive  alone  
275  woman       False    D  Southampton   yes  False  
483  woman       False  NaN  Southampton   yes   True  
829  woman       False    B          NaN   yes   True  
366  woman       False    D    Cherbourg   yes  False  
268  woman       False    C  Southampton   yes  False

### Exercise 6

Find the record of the oldest female who did not survive.

**Check:** She was 57, embarked in Southampton and traveled in the second class. 

In [87]:
# "nbsphinx": "hidden"

df[(df["sex"]=="female") & (df["survived"]==0)].sort_values(by="age", ascending=False).head(1)

survived  pclass     sex   age  sibsp  parch  fare embarked   class  \
772         0       2  female  57.0      0      0  10.5        S  Second   

       who  adult_male deck  embark_town alive  alone  
772  woman       False    E  Southampton    no   True

### Exercise 7

Find the number of people who survived the sinking and the number of people who died.

**Check:** 342 survived, 549 died. 

In [53]:
# "nbsphinx": "hidden"

df["survived"].value_counts()

0    549
1    342
Name: survived, dtype: int64

### Exercise 8

What was the average age of people who survived?

**Check:** 28.34 years. 

In [71]:
# "nbsphinx": "hidden"

df[df["survived"]==1]["age"].mean()

28.343689655172415

### Exercise 9

There were three classes of passengers aboard Titanic: "First", "Second" and "Third". Compute what fraction of passengers traveling in each class survived.

**Check:** First: 0.629, Second: 0.473, Third: 0.242.

In [78]:
# "nbsphinx": "hidden"

classes = ["First", "Second", "Third"]
for cl in classes:
    df_cl = df[df["class"]==cl]
    print(f"{cl}: {len(df_cl[df_cl['survived']==1])/len(df_cl)}")

First: 0.6296296296296297
Second: 0.47282608695652173
Third: 0.24236252545824846
